### ISSUE TITLE
ResizeObserver not available on macOS 12

### ISSUE THREAD

**Issue by purfectliterature** [SafeExamBrowser/seb-mac#335]
On macOS 12.0.1 (at least), [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) is not available in SEB (at least version 3.2.5). Websites using ResizeObserver will raise a `ReferenceError` if not polyfilled. macOS 12.0.1 ships with Safari 15, and it already has ResizeObserver available.

SEB on macOS 13 and later has ResizeObserver available.

I understand that SEB on macOS is using WKWebView, and that means that it runs WebKit that is bundled with the host macOS. This means that some Web APIs may or may not be available across clients with different macOS versions. But this behaviour is uber-weird since ResizeObserver should be available already.

---

On a separate note, SEB's WKWebView API parity is so weird. WKWebView should follow the host macOS's WebKit version. So, on macOS 13 (with Safari 16), it should support [SharedWorker](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker) since [the support was added in Safari 16](https://webkit.org/blog/13152/webkit-features-in-safari-16-0/#:~:text=of%20the%20page.-,Shared%20Workers,-WebKit%20now%20supports). But SEB doesn't.

I tried to create my own Mac app with WKWebView, and it follows my intuition. ResizeObserver available on macOS 12. SharedWorker available on macOS 13 and later. Apart from Service Worker and other things that require the [`web-browser` entitlement](https://developer.apple.com/documentation/xcode/preparing-your-app-to-be-the-default-browser), everything is pretty much similar to the host's Safari.

Any idea as to why this is the case with SEB on macOS? I tried skimming through SEBOSXWKWebViewController and SEBAbstractModernWebView, and still couldn't figure out why SEB's WKWebView behaves differently from my plain WKWebView app.

I'm trying to figure out how SEB works on macOS because I'm developing an LMS that is used for practical exams at the National University of Singapore. We very much want to reduce possibilities of crashes during the exams.

Thanks for your help!

**Comment by danschlet** [comment]
Are you sure that SEB is using WKWebView ("modern Webview") with yours SEB settings? Depending on settings, SEB can still use the classic WebView (UIWebView on iOS) for compatibiltiy reasons (as sending the BrowserExamKey and ConfigKey in request headers isn't possible with WKWebView). Unless you disabled zooming in your SEB settings, you can find out if WKWebView is used by trying pinch-to-zoom, which only works in the modern, not the classic WebView. You can also check the SEB log if it contains the string "Opening modern WebView".

Nice to hear that NUS is using (or plans to) SEB! Greetings to Singapore!

**Comment by purfectliterature** [comment]
Gosh, that did it! I knew that SEB has two browser flavours, but I thought we were using the modern one all along! We just went with the default _Prefer Modern in New Tab+Different Host_, until I saw your reply, and reread that setting, and noticed __in New Tab__ which means the main tab is using the classic browser!

I verified the case by switching to _Prefer Modern_ and now ResizeObserver is defined and the log says `Opening modern WebView`. Thanks so much!

I am already using the JavaScript API for collecting the BEK and Config Key, so the no-keys-in-request-header is a non-issue for me. But apart from this, any other reasons why one might not want to switch to _Prefer Modern_? Or rather, SEB having it set to that by default?

NUS (the School of Computing, at least) is using it for the second year in our programming fundamental courses, actually. So, it's no longer a plan. Thanks so much for developing this thing so we can upkeep the integrity of our tests!

---

On another note, the developer tools in WKWebView flickers on SEB. I think the WKWebView needs to have Autoresizing Mask set on it. It fixed the flickers on my own plain WKWebView test app!

**Comment by danschlet** [comment]
> I am already using the JavaScript API for collecting the BEK and Config Key, so the no-keys-in-request-header is a non-issue for me. But apart from this, any other reasons why one might not want to switch to _Prefer Modern_? Or rather, SEB having it set to that by default?

As with WKWebView the BEK/CK cannot be send in request headers, with the setting sendBrowserExamKey=true for backwards compatibility the classic WebView is used (as described [here](https://safeexambrowser.org/developer/seb-config-key.html)). That's why you explicitely have to set browserWindowWebView=3 (prefer modern) to signal to SEB that you're supporting the JS API and don't need BEK/CK in headers. If you would set sendBrowserExamKey=false, SEB would anyways use WKWebView by default. That wasn't advisable before SEB for Windows (3.3.2) supported the JS API, now the sendBrowserExamKey=true setting isn't necessary anymore.

I will also add URL content filter support when using WKWebView soon, as currently setting URLFilterEnableContentFilter=true would also force SEB to use the classic WebView (regardless of the settings mentioned above).


> NUS (the School of Computing, at least) is using it for the second year in our programming fundamental courses, actually. So, it's no longer a plan—it's actual! Thanks so much for developing this thing so we can upkeep the integrity of our tests!

Welcome, it's always satisfying to see that our software is appreciated around the world (and I have family in Singapore and a relative did his PhD at NUS ;-) )!

> On another note, the developer tools in WKWebView flickers on SEB. I think the WKWebView needs to have Autoresizing Mask set on it. It fixed the flickers on my own plain WKWebView test app!

Wow, that's great news! I was convinced that WebInspector doesn't work properly with WKWebView because Apple doesn't officially support it. To what values did you set Autoresizing Mask exactly? I'm using Auto Layout Constraints, maybe that's why it isn't displayed correctly: https://github.com/SafeExamBrowser/seb-mac/blob/fbdf8b7e71bfbf163a10f886a9461578bf9022e3/Classes/BrowserComponents/SEBOSXWKWebViewController.swift#L61

**Comment by purfectliterature** [comment]
Seems like the classic WebView is there to support LMSes that still require BEK/CK over HTTP request headers. Then I think we're in good shape to use the modern WKWebView because we are using the SEB JavaScript API to retrieve the BEK/CK. Thanks for your clarifications!

>Welcome, it's always satisfying to see that our software is appreciated around the world (and I have family in Singapore and a relative did his PhD at NUS ;-) )!

Cool, cheers!

>To what values did you set Autoresizing Mask exactly?

I'm actually using the Interface Builder to add a WKWebView to the storyboard, so I bound it via an IBOutlet. Initially, I was added constraints of `0`s on top, bottom, leading, and trailing but it causes flickers in the Web Inspector. So, [I found this thread](https://stackoverflow.com/questions/60727065/wkwebview-web-inspector-in-macos-app-fails-to-render-and-flickers-flashes), and I switched the WKWebView's Layout on the Size Inspector to Autoresizing Mask and set it like this.

![image](https://github.com/SafeExamBrowser/seb-mac/assets/51525686/cdc37785-5237-4533-bdff-b6036684cdd6)

No more flickers, and my WKWebView still resizes with the window!

But this is equivalent to what you have on the line just above the snippet you sent me. So, I'm not sure why it still flickers on SEB...

https://github.com/SafeExamBrowser/seb-mac/blob/fbdf8b7e71bfbf163a10f886a9461578bf9022e3/Classes/BrowserComponents/SEBOSXWKWebViewController.swift#L60C1-L60C1

Although I didn't have to translate my Autoresizing Mask to constraints. Seems like it's defaulting to `true`, but setting `translatesAutoresizingMaskIntoConstraints` to `false` didn't change anything for me.

Though, my WKWebView's superview `resizesSubviews`.
